src/apps/weblib/ts-api/data.ts

```diff
@@ -82,15 +82,42 @@ export class Data {
 
 		const values = series.values ? series.values : ([] as D.Values)
 
-		const seriesType = series.type ? series.type : this._detectType(values)
+		if ('categories' in series) {
+			if (!Array.isArray(series.categories)) {
+				throw new Error('categories field is not an array')
+			}
+			this._checkCategories(values, series.categories)
+			this._addDimension(series.name, values as number[], series.categories)
+		}
+		else {
+			const seriesType = series.type ? series.type : this._detectType(values)
+			if (seriesType === 'dimension') {
+				const { indexes, categories } = this._convertDimension(values)
+				this._addDimension(series.name, indexes, categories)
+			} else if (seriesType === 'measure') {
+				if (!series.unit) series.unit = ''
+				this._addMeasure(series.name, series.unit, values)
+			} else {
+				throw new Error('invalid series type: ' + series.type)
+			}
+		}
+	}
 
-		if (seriesType === 'dimension') {
-			this._addDimension(series.name, values)
-		} else if (seriesType === 'measure') {
-			if (!series.unit) series.unit = ''
-			this._addMeasure(series.name, series.unit, values)
-		} else {
-			throw new Error('invalid series type: ' + series.type)
+	private _checkCategories(values: (string | number | null)[], categories: (string | number | null)[]): void {
+		if (!values.every((value) => value === null || Number.isInteger(value))) {
```

Comment:
What happens when a value is null? We must not let it add a null value to the values array.